### PR TITLE
Dat 3521 mercury empty sections headers removal

### PR DESCRIPTION
**Pull request by idradm**

### CODE REVIEW

File: includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
+<?php
+
+class TemplateCleanup {
```
Comment (sebastianmarzjan) [inline includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php]: I'm not sure if the name covers what this class is responsible for - maybe DocumentCleanup or HtmlCleanup would be more informative

File: includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	 * @return bool
+	 */
+	public static function doCleanup( $parser, &$html ) {
+		wfProfileIn( __METHOD__ );
+
+		if ( $html && static::shouldBeProcessed() ) {
+			$dom = HtmlHelper::createDOMDocumentFromText( $html );
+
+			$xpath = new DOMXPath( $dom );
+			$headers = [ 'h6', 'h5', 'h4', 'h3', 'h2', 'h1' ];
```
Comment (sebastianmarzjan) [inline includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php]: What do you think about making this a class field? It'd be easier to see the definitions right away that way.

File: includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		}
+		wfProfileOut( __METHOD__ );
+
+		return true;
+	}
+
+	private static function shouldBeProcessed() {
+		global $wgEnableEmptySectionsCleanup, $wgArticleAsJson;
+
+		return $wgEnableEmptySectionsCleanup && $wgArticleAsJson;
```
Comment (sebastianmarzjan) [inline includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php]: Is there a corresponding PR for the config repo defining the default?

File: includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+	/**
+	 * This method will remove empty sections headers from article content
+	 *
+	 * @param $parser
+	 * @param $html
+	 *
+	 * @return bool
+	 */
+	public static function doCleanup( $parser, &$html ) {
```
Comment (sebastianmarzjan) [inline includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php]: One thing I noticed about this method's output is that it changes whitespace in lists:

```
<ul><li> 1
</li><li> 2
</li><li> 3
</li></ul>
```

becomes

```
<ul>
<li> 1
</li>
<li> 2
</li>
<li> 3
</li>
</ul>
```

I do not think it's posing an issue in our case; What's your opinion?

File: includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+		wfProfileIn( __METHOD__ );
+
+		if ( $html && static::shouldBeProcessed() ) {
+			$dom = HtmlHelper::createDOMDocumentFromText( $html );
+
+			$xpath = new DOMXPath( $dom );
+			$headers = [ 'h6', 'h5', 'h4', 'h3', 'h2', 'h1' ];
+			foreach ( $headers as $h ) {
+				$hs = $xpath->query( "//{$h}" );
```
Comment (sebastianmarzjan) [inline includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php]: $headerNodes / $headerNodeList? :)

File: includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+					$current = $hs->item( $i );
+					//skip empty text nodes (spaces and new lines)
+					$emptySiblings = [ ];
+					$next = $current->nextSibling;
+					while ( $next && $next->nodeType == XML_TEXT_NODE && empty( trim( $next->nodeValue ) ) ) {
+						$emptySiblings[] = $next;
+						$next = $next->nextSibling;
+					}
+					//remove if next node is header of same or higher level, or last node
+					if ( in_array( $next->nodeName, $headers ) && $current->nodeName[ 1 ] >= $next->nodeName[ 1 ]
```
Comment (sebastianmarzjan) [inline includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php]: What do you think about adding parenthesis to make the evaluation order explicit?

File: includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+				for ( $i = 0; $i < $hs->length; $i++ ) {
+					$current = $hs->item( $i );
+					//skip empty text nodes (spaces and new lines)
+					$emptySiblings = [ ];
+					$next = $current->nextSibling;
+					while ( $next && $next->nodeType == XML_TEXT_NODE && empty( trim( $next->nodeValue ) ) ) {
+						$emptySiblings[] = $next;
+						$next = $next->nextSibling;
+					}
+					//remove if next node is header of same or higher level, or last node
```
Comment (sebastianmarzjan) [inline includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php]: Love the comments :heart:

File: includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+					if ( in_array( $next->nodeName, $headers ) && $current->nodeName[ 1 ] >= $next->nodeName[ 1 ]
+						 || $next === null
+					) {
+						$current->parentNode->removeChild( $current );
+						//remove empty siblings as well
```
Comment (sebastianmarzjan) [inline includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php]: L37-41 could go into separate method ( i.e. removeNodes($emptySiblings) ) to shorten this function

File: includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+		if ( $html && static::shouldBeProcessed() ) {
+			$dom = HtmlHelper::createDOMDocumentFromText( $html );
+
+			$xpath = new DOMXPath( $dom );
+			$headers = [ 'h6', 'h5', 'h4', 'h3', 'h2', 'h1' ];
+			foreach ( $headers as $h ) {
+				$hs = $xpath->query( "//{$h}" );
+				for ( $i = 0; $i < $hs->length; $i++ ) {
+					$current = $hs->item( $i );
+					//skip empty text nodes (spaces and new lines)
```
Comment (sebastianmarzjan) [inline includes/wikia/parser/templatetypes/handlers/TemplateCleanup.class.php]: I know it'd give us two return values; Still, I am wondering what would you say about extracting L25-31 along the lines of 

```
list( $nextNode, $emptySiblings ) = $this->skipEmptySiblings($current);
```

File: includes/wikia/parser/templatetypes/TemplateTypes.setup.php [hunk 10]
```diff
@@ -1,30 +1,35 @@
...
-$wgAutoloadClasses['NavigationTemplate'] = __DIR__ . '/handlers/NavigationTemplate.class.php';
+$wgAutoloadClasses[ 'ContextLinkTemplate' ] = __DIR__ . '/handlers/ContextLinkTemplate.class.php';
+$wgAutoloadClasses[ 'InfoIconTemplate' ] = __DIR__ . '/handlers/InfoIconTemplate.class.php';
+$wgAutoloadClasses[ 'NavboxTemplate' ] = __DIR__ . '/handlers/NavboxTemplate.class.php';
+$wgAutoloadClasses[ 'NoticeTemplate' ] = __DIR__ . '/handlers/NoticeTemplate.class.php';
+$wgAutoloadClasses[ 'QuoteTemplate' ] = __DIR__ . '/handlers/QuoteTemplate.class.php';
+$wgAutoloadClasses[ 'ReferencesTemplate' ] = __DIR__ . '/handlers/ReferencesTemplate.class.php';
+$wgAutoloadClasses[ 'ScrollboxTemplate' ] = __DIR__ . '/handlers/ScrollboxTemplate.class.php';
+$wgAutoloadClasses[ 'NavigationTemplate' ] = __DIR__ . '/handlers/NavigationTemplate.class.php';
+$wgAutoloadClasses[ 'TemplateCleanup' ] = __DIR__ . '/handlers/TemplateCleanup.class.php';
```
Comment (sebastianmarzjan) [inline includes/wikia/parser/templatetypes/TemplateTypes.setup.php]: Nitpick - per comment about the class name, it makes sense to me to move it below DataTables

**Comment by sebastianmarzjan** [comment]
What I really like about this code is that it removes all the stub headers, not only those coming from templates :)

**Comment by idradm** [comment]
Config changes: https://github.com/Wikia/config/pull/1525

**Comment by sebastianmarzjan** [comment]
LGTM.
Do you want to add support for paragraphs / breaklines here too?

**Comment by idradm** [comment]
Not yet, let's wait and see if it's an often case.

**Comment by sebastianmarzjan** [comment]
:+1: LGTM :+1:

**Pull request merged**